pkg/storage/openstack.go

```diff
@@ -0,0 +1,182 @@
+package storage
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	pathutil "path"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	osObjects "github.com/gophercloud/gophercloud/openstack/objectstorage/v1/objects"
+)
+
+// ReauthRoundTripper satisfies the http.RoundTripper interface and is used to
+// limit the number of re-auth attempts (infinite by default)
+type ReauthRoundTripper struct {
+	rt                http.RoundTripper
+	numReauthAttempts int
+}
+
+// RoundTrip performs a round-trip HTTP request and logs relevant information about it.
+func (rrt *ReauthRoundTripper) RoundTrip(request *http.Request) (*http.Response, error) {
+	response, err := rrt.rt.RoundTrip(request)
+	if response == nil {
+		return nil, err
+	}
+
+	if response.StatusCode == http.StatusUnauthorized {
+		if rrt.numReauthAttempts == 3 {
+			return response, errors.New("tried to re-authenticate 3 times with no success")
+		}
+		rrt.numReauthAttempts++
+	}
+
+	return response, nil
+}
+
+// OpenstackOSBackend is a storage backend for Openstack Object Storage
+type OpenstackOSBackend struct {
+	Container string
+	Prefix    string
+	Region    string
+	CACert    string
+	Client    *gophercloud.ServiceClient
+}
+
+// NewOpenstackOSBackend creates a new instance of OpenstackOSBackend
+func NewOpenstackOSBackend(container string, prefix string, region string, caCert string) *OpenstackOSBackend {
+	authOptions, err := openstack.AuthOptionsFromEnv()
+	if err != nil {
+		panic(fmt.Sprintf("Openstack (environment): %s", err))
+	}
+	authOptions.AllowReauth = true
+
+	// Create a custom HTTP client to handle reauth retry and custom CACERT if needed
+	roundTripper := ReauthRoundTripper{}
+	if caCert != "" {
+		caCert, err := ioutil.ReadFile(caCert)
+		if err != nil {
+			panic(fmt.Sprintf("Openstack (ca certificates): %s", err))
+		}
+
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
```

Comment:
Should we be droping the return value here ?

Reply:
Yeah, even if it does not return an error, I can throw a pem parsing error or the like myself indeed.